torch_glow/src/PyTorchModelLoader.cpp

```diff
@@ -25,47 +28,354 @@ glow::Type ptTypeToGlowType(const at::CompleteTensorType &ptType) {
+  assert(handle.size() == 1 ||
+         handle.size() == targetSize &&
+             "Expected input size to be either 1 or the target size");
+  std::vector<OutT> out;
+  out.resize(targetSize);
+  for (size_t i = 0; i < targetSize; ++i) {
+    out[i] =
+        static_cast<OutT>(handle.size() == 1 ? handle.raw(0) : handle.raw(i));
+  }
+  return out;
+}
+
+/// Given a Glow Handle \p Handle returns the first element in the Tensor
+/// of that Handle, asserting it is equivalent to all other elements in the
+/// Tensor.
+template <typename T> T contractParamIfNeeded(const glow::Handle<T> &handle) {
+  assert(handle.size() > 0);
+  const T firstElem = handle.raw(0);
+  for (size_t i = 1; i < handle.size(); i++) {
+    assert(handle.raw(i) == firstElem);
+  }
+  return firstElem;
+}
 } // namespace
 
-glow::Placeholder *PyTorchModelLoader::loadValue(const torch::jit::Value *val) {
-  // TODO: do we need to care about optional IValues here?
-  assert(val->isCompleteTensor());
-  auto ptType = val->type()->cast<at::CompleteTensorType>();
+glow::NodeValue
+PyTorchModelLoader::getGlowNodeValue(const torch::jit::Value *value) const {
+  const auto it = valueMap_.find(value);
+  assert(it != valueMap_.end());
+  return it->second;
+}
+
+bool PyTorchModelLoader::hasGlowNodeValue(
+    const torch::jit::Value *value) const {
+  const auto it = valueMap_.find(value);
+  return it != valueMap_.end();
+}
+
+void PyTorchModelLoader::addGlowNodeValue(const torch::jit::Value *value,
+                                          glow::NodeValue nodeValue) {
+  assert(valueMap_.count(value) == 0);
+  valueMap_[value] = nodeValue;
+}
+
+template <typename T>
+glow::Handle<T> PyTorchModelLoader::getGlowConstantHandle(
+    const torch::jit::Value *value) const {
+  glow::NodeValue nodeValue = getGlowNodeValue(value);
+  return getGlowConstantPayload<T>(nodeValue.getNode());
+}
+
+glow::Placeholder *
+PyTorchModelLoader::loadValue(const torch::jit::Value *value) {
+  assert(value->isCompleteTensor());
+  auto ptType = value->type()->cast<at::CompleteTensorType>();
   auto glowType = ptTypeToGlowType(*ptType.get());
   glow::Placeholder *ph =
       mod_.createPlaceholder(&glowType, "input", /*isTrainable*/ false);
-  valueMap_[val] = ph->getOutput();
+  addGlowNodeValue(value, ph->getOutput());
   return ph;
 }
 
-void PyTorchModelLoader::loadNode(const torch::jit::Node *ptNode) {
-  glow::Node *glowNode = nullptr;
+void PyTorchModelLoader::loadMul(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 2);
+  assert(outputs.size() == 1);
+
+  glow::NodeValue rhs = getGlowNodeValue(inputs[0]);
+  glow::NodeValue lhs = getGlowNodeValue(inputs[1]);
+  glow::MulNode *glowNode = f_->createMul("mul", rhs, lhs);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
 
+void PyTorchModelLoader::loadDiv(const torch::jit::Node *ptNode) {
```

Comment:
Great! I think this kind of design is much better than switch case!

Reply:
Agreed, a while ago I converted the Caffe2 model loader to this style, I think it's easier to work with.